### ISSUE TITLE
Performance considerations of using multiple (nested) buckets?

### ISSUE THREAD

**Issue by Xeoncross** [etcd-io/bbolt#120]
Considering buckets can [be used to namespace objects](https://github.com/etcd-io/bbolt#using-buckets) for security, are their performance benefits to using buckets instead of longer key prefixes?

```
b := tx.Bucket("app")
b = b.Bucket("user")
b.Get("12:name")
```
vs
```
b := tx.Bucket("app")
b.Get("user:12:name")
```

**Comment by Xeoncross** [comment]
I went ahead and just created a database using a single bucket and one using a nested bucket (as shown above) and for 100k records in `{id}:{prop_name}` format I found that using the nested bucket required less space. This is probably because of the way the bucket name is stored resulting in smaller prefixes to all keys stored inside it (just guessing).

```
Flat ("user:{id}:{prop_name}" format)
Data size: 21135019
File size: 74289152
&{Data:69255168 PageSize:4096}

Nested "user" Bucket
Data size: 19132730
File size: 69922816
&{Data:78299136 PageSize:4096}
```
Reducing the size of the bucket name down to 1 byte brought the results closer together.

```
Flat with []byte("u")
Data size: 19527365
File size: 70766592
&{Data:84590592 PageSize:4096}
```

**Comment by Xeoncross** [comment]
Found an ticket on the original boltdb repo [about using nested buckets vs namespacing](https://github.com/boltdb/bolt/issues/422). 

> If you have a large number of keys in the sub-bucket then you can save space by not having to prefix each one. Bolt doesn't do key compression like LSMs do. However, if you have a small number of keys then you'll likely have better performance with simple prefixing since Bolt won't have to traverse down another level into the sub-bucket's b-tree.

If it's true that each bucket has it's own b-tree, then it would seem you could get better write performance by segregating every type of object into it's own bucket reducing the size of each objects b-tree that had to be restructured on insert.

**Comment by kylewolfe** [comment]
My question on this on the original repo: boltdb/bolt#473

**Comment by cenkalti** [comment]
Summarizing above information.

Using nested buckets vs prefixing keys is a trade off.

Advantages of nested-bucket:
- You can omit the prefix and make key size smaller

Disadvantages of nested-bucket:
- bbolt first needs to find the bucket by traversing the top-level bucket. Then it needs to traverse the nested bucket to find the key position. Each nesting level add additional search.

Advantages of prefixing:
- No need to traverse additional tree for sub-bucket

Disadvantages of prefixing:
- Key size is larger because of adding prefix to each key

**Comment by itsfarseen** [comment]
Here are some concrete numbers on how nested buckets affect performance:
https://github.com/itsfarseen/boltdb-benchmarks

TLDR; 
- Both nested buckets and prefixed keys take a huge hit on write performance as the number of records increase. They're slower than serializing and writing JSON, one KV pair per record.
- Both nested buckets and prefixed keys are faster than JSON, but slower than storing a record manually marshaleld into binary in a single KV pair, when reading a single record.
- Prefixed keys are as fast as manual binary encoding when reading meny records in a single transaction. Nested buckets are slightly slower, but faster than JSON.
- Nested buckets, prefixed keys and manual binary encoding perform the same, and are the best, when reading/updating a single field.
